source/common/upstream/original_dst_cluster.cc

```diff
@@ -0,0 +1,163 @@
+          if (localIP->version() == Network::Address::IpVersion::v4) {
+            hostIpPort.reset(
+                new Network::Address::Ipv4Instance(localIP->addressAsString(), localIP->port()));
+          } else {
+            hostIpPort.reset(
+                new Network::Address::Ipv6Instance(localIP->addressAsString(), localIP->port()));
+          }
+          HostSharedPtr host = parent_.createHost(hostIpPort);
+
+          ENVOY_LOG(debug, "CREATED HOST {}", host->address()->asString());
+          // Add the new host to the map.  We just failed to find it in
+          // our local map above, so just insert it.
+          host_map_.insert({localAddr.asString(), host});
+          host->weight(2); // Mark as used.
+          return host;
+        } else {
+          ENVOY_LOG(debug, "FAILED TO CREATE HOST FOR {}", localAddr.asString());
+        }
+      }
+    }
+  }
+
+  ENVOY_LOG(warn, "original_dst_load_balancer: No downstream connection or no original_dst.");
+  return nullptr;
+}
+
+OriginalDstCluster::OriginalDstCluster(const Json::Object& config, Runtime::Loader& runtime,
+                                       Stats::Store& stats,
+                                       Ssl::ContextManager& ssl_context_manager,
+                                       Event::Dispatcher& dispatcher, bool added_via_api)
+    : ClusterImplBase(config, runtime, stats, ssl_context_manager, added_via_api),
+      dispatcher_(dispatcher), cleanup_interval_ms_(std::chrono::milliseconds(
+                                   config.getInteger("cleanup_interval_ms", 5000))),
+      cleanup_timer_(dispatcher.createTimer([this]() -> void { cleanup(); })) {
+  size_t n_hosts = 0;
+  try {
+    std::vector<Json::ObjectSharedPtr> hosts_json = config.getObjectArray("hosts");
+    n_hosts = hosts_json.size();
+  } catch (...) {
+  } // Catch missing hosts config.
+  if (n_hosts != 0) {
+    throw EnvoyException("original_dst clusters must have no hosts configured");
+  }
+
+  cleanup_timer_->enableTimer(cleanup_interval_ms_);
+}
+
+// Called by a worker thread.
+HostSharedPtr OriginalDstCluster::createHost(Network::Address::InstanceConstSharedPtr address) {
+  HostSharedPtr host(
+      new HostImpl(info_, info_->name() + address->asString(), address, false, 1, ""));
+  dispatcher_.post([this, host]() -> void { addHost(host); });
+  return host;
+}
+
+// Called in the main thread.
+void OriginalDstCluster::addHost(HostSharedPtr host) {
+  HostVectorSharedPtr new_hosts(new std::vector<HostSharedPtr>(hosts()));
+  new_hosts->emplace_back(host);
+  updateHosts(new_hosts, createHealthyHostList(*new_hosts), empty_host_lists_, empty_host_lists_,
+              {host}, {});
+}
+
+void OriginalDstCluster::cleanup() {
+  ENVOY_LOG(debug, "CLEANUP.");
+  HostVectorSharedPtr new_hosts(new std::vector<HostSharedPtr>);
+  std::vector<HostSharedPtr> to_be_removed;
+  auto hostset = hosts();
+  for (auto host : hostset) {
+    if (host->weight() == 1) {
```

Comment:
I see you are somehow using weight for cleanup marking. Can you comment/explain that somewhere?